torch/_dynamo/eval_frame.py

```diff
@@ -188,6 +270,34 @@ def _fn(*args, **kwargs):
         # If the function is called using torch._dynamo.optimize decorator, we
         # should prevent any type of skipping.
         if callback not in (None, False):
+            if not hasattr(fn, "__code__"):
+                raise RuntimeError(
+                    textwrap.dedent(
+                        """
+
+                        torch._dynamo.optimize is called on a non function object.
+                        If this is a callable class, please optimize the individual methods that you are interested in optimizing.
```

Comment:
Do we want to do this? It feels like a significant step down in UX from where we started i.e: the context manager

Reply:
I'm not sure how this is any worse from what we have now?

Reply:
I'm thinking of it as the difference between

`optimize(mod.forward)` and 

`def wrapper(): mod; optimize(wrapper)`